rasterio/rio/shapes.py

```diff
@@ -110,134 +106,147 @@ def shapes(
-
-                # Transform the raster bounds.
-                bounds = src.bounds
-                xs = [bounds[0], bounds[2]]
-                ys = [bounds[1], bounds[3]]
-                if projection == 'geographic':
-                    xs, ys = rasterio.warp.transform(
-                        src.crs, CRS({'init': 'epsg:4326'}), xs, ys)
-                if precision >= 0:
-                    xs = [round(v, precision) for v in xs]
-                    ys = [round(v, precision) for v in ys]
-                self._xs = xs
-                self._ys = ys
-
-                # Prepare keyword arguments for shapes().
-                kwargs = {'transform': transform}
-                if not with_nodata:
-                    kwargs['mask'] = msk
-
-                src_basename = os.path.basename(src.name)
-
-                # Yield GeoJSON features.
-                for i, (g, val) in enumerate(
-                        rasterio.features.shapes(img, **kwargs)):
-                    if projection == 'geographic':
-                        g = rasterio.warp.transform_geom(
-                            src.crs, 'EPSG:4326', g,
-                            antimeridian_cutting=True, precision=precision)
-                    xs, ys = zip(*coords(g))
-                    yield {
-                        'type': 'Feature',
-                        'id': "{0}:{1}".format(src_basename, i),
-                        'properties': {
-                            'val': val, 'filename': src_basename
-                        },
-                        'bbox': [min(xs), min(ys), max(xs), max(ys)],
-                        'geometry': g
-                    }
-
     if not sequence:
         geojson_type = 'collection'
 
     try:
         with ctx.obj['env'] as env:
             write_features(
-                stdout, Collection(env), sequence=sequence,
+                stdout,
+                feature_gen(
+                    input, env, bidx, sampling,  band, as_mask, with_nodata, projection, precision),
+                sequence=sequence,
                 geojson_type=geojson_type, use_rs=use_rs,
                 **dump_kwds)
     except Exception:
         logger.exception("Exception caught during processing")
         raise click.Abort()
+
+
+def feature_gen(input, env, *args, **kwargs):
+    class Collection(object):
+
+        def __init__(self, env):
+            self.bboxes = []
+            self.env = env
+
+        @property
+        def bbox(self):
+            minxs, minys, maxxs, maxys = zip(*self.bboxes)
+            return min(minxs), min(minys), max(maxxs), max(maxys)
+
+        def __call__(self):
+            for f in features(input, *args, **kwargs):
+                self.bboxes.append(f['bbox'])
+                yield f
+
+    return Collection(env)
+
+
+def features(input, bidx, sampling, band, as_mask, with_nodata, projection, precision):
+    import rasterio
+    import numpy as np
+
+    with rasterio.open(input) as src:
```

Comment:
@perrygeo if `features` took `src` as input instead of the path, we wouldn't need to add more args to `features` if we wanted to use esoteric GDAL open options. I'm thinking along the lines of http://python-patterns.guide/gang-of-four/factory-method/#dodge-use-dependency-injection here.